tests/mir_test_framework/window_management_test_harness.cpp

```diff
@@ -0,0 +1,213 @@
+/*
+ * Copyright © Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 or 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "mir_test_framework/window_management_test_harness.h"
+#include <mir/executor.h>
+#include <mir/scene/surface.h>
+#include <mir/shell/surface_specification.h>
+#include <mir/wayland/weak.h>
+#include <miral/window_manager_tools.h>
+#include <miral/minimal_window_manager.h>
+#include <miral/set_window_management_policy.h>
+#include "mir/test/doubles/fake_display.h"
+#include "mir/test/doubles/stub_buffer_stream.h"
+#include "src/miral/basic_window_manager.h"
+
+#include <queue>
+
+namespace ms = mir::scene;
+namespace msh = mir::shell;
+namespace mf = mir::frontend;
+namespace mi = mir::input;
+namespace mtd = mir::test::doubles;
+namespace mg = mir::graphics;
+using namespace testing;
+
+namespace
+{
+class WorkQueue
+{
+public:
+    void enqueue(std::function<void()> const& f)
+    {
+        work_queue.push(f);
+    }
+
+    auto next() -> bool
+    {
+        if (work_queue.empty())
+            return false;
+
+        auto const& f = work_queue.front();
+        f();
+        work_queue.pop();
+        return !work_queue.empty();
+    }
+
+    void run()
+    {
+        while (next()) {}
+    }
+
+private:
+    std::queue<std::function<void()>> work_queue;
+};
```

Comment:
I don't think you need the `WorkQueue` anymore? There's no longer recursion into multiple observers from one observation.

Reply:
We still need the `WorkQueue` because there will be recursion with event handling, for example.